File: BFS/BFS.java
import java.util.*;

class Graph
{
	private int v; // vertices
	private LinkedList<Integer> adj[]; //Adjacency List

	// Constructor
	Graph(int n)
	{
		v = n;
		adj = new LinkedList[v];
		for (int i=0; i<v; ++i) {
			adj[i] = new LinkedList();
		}
	}
	//Add an edge to graph
	void addEdge(int v,int w)
	{
		adj[v].add(w);
	}

	//BFS traversal
	void BFS(int s)
	{
		boolean vis[] = new boolean[v];

		// Create a queue for BFS
		LinkedList<Integer> q = new LinkedList<Integer>();
		
		//Set the node as visited and enqueue
		vis[s]=true;
		q.add(s);

		while (q.size() != 0)
		{
			// Dequeue a vertex and print 
			s = q.poll();
			System.out.print(s+" ");

			Iterator<Integer> i = adj[s].listIterator();
			while (i.hasNext())
			{
				int n = i.next();
				if (!vis[n])
				{
					vis[n] = true;
					q.add(n);
				}
			}
		}
	}
}

//Time Complexity: O(v+e)
//Auxiliary Space: O(v) 

//v = number of vertices and e = number of edges.

public class BFS {
	public static void main(String args[])
	{
		Graph graph = new Graph(4);

		graph.addEdge(0, 1);
		graph.addEdge(0, 2);
		graph.addEdge(1, 2);
		graph.addEdge(2, 0);
		graph.addEdge(2, 3);
		graph.addEdge(3, 3);

		System.out.println("BFS Traversal : ");
		graph.BFS(2);
	}
}

File: algorithms/BFS/bfs.ts
import Queue from "./queue";

class Graph<T> {
    list: Map<T, Array<T>>;
  
    constructor() {
      this.list = new Map<T, Array<T>>();
    }
  
    addVertice(value: T) {
      this.list.set(value, []);
    }
  
    addEdge(src: T, dest: T) {
      this.list.get(src).push(dest);
      this.list.get(dest).push(src);
    }
  
    bfs(node, callback) {
      let q = new Queue<T>();
  
      let visited = new Set<T>();
  
      q.enqueue(node);
      visited.add(node);
  
      while (!q.isEmpty()) {
        let element = q.dequeue();
  
        callback(element);
  
        let items = this.list.get(element);
        for (let i in items) {
          let neighbor = items[i];
          if (!visited.has(neighbor)) {
            q.enqueue(neighbor);
            visited.add(neighbor);
          }
        }
      }
    }
}